File: 0104-maximum-depth-of-binary-tree/0104-maximum-depth-of-binary-tree.rb
# Definition for a binary tree node.
# class TreeNode
#   attr_accessor :val, :left, :right
#   def initialize(val = 0, left = nil, right = nil)
#     @val = val
#     @left = left
#     @right = right
#   end
# end
# @param {TreeNode} root
# @return {Integer}
def max_depth(root)
  return 0 if !root
  depth = 0
  levels = Hash.new{ |h, k| h[k] = Array.new }
  levels[depth] = [root]
  while levels[depth].length > 0
    levels[depth].each do |node|
      if node.left
        levels[depth+1] << node.left
      end
      if node.right
        levels[depth+1] << node.right
      end
    end
    depth += 1
  end
  return depth
end

File: 0104-maximum-depth-of-binary-tree/0104-maximum-depth-of-binary-tree.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *   Val int
 *   Left *TreeNode
 *   Right *TreeNode
 * }
 */
func maxDepth(root *TreeNode) int {
  if root == nil {
    return 0
  }
  depth, levels := 0, [][]*TreeNode{[]*TreeNode{root}}
  for len(levels[depth]) > 0 {
    if len(levels) < depth + 2 {
      levels = append(levels, []*TreeNode{})
    }
    for _, node := range levels[depth] {
      if (*node).Left != nil {
        levels[depth + 1] = append(levels[depth + 1], (*node).Left)
      }
      if (*node).Right != nil {
        levels[depth + 1] = append(levels[depth + 1], (*node).Right)
      }
    }
    depth++
  }
  return depth
}